versioned/storage/common/src/test/java/org/projectnessie/versioned/storage/common/logic/TestMergeBase.java

```diff
@@ -0,0 +1,479 @@
+                .targetCommitId(e.id())
+                .fromCommitId(f.id())
+                .build()
+                .identifyMergeBase())
+        .isEqualTo(a.id());
+  }
+
+  /**
+   * Merge-again case. <code><pre>
+   *              ----E----G---------I
+   *            /       /           /
+   *       ----B----D----------H----
+   *      /               /
+   * ----A------C--------F
+   * </pre></code>
+   *
+   * <p>Best merge-base of {@code I} onto {@code F} is {@code F}.
+   */
+  @Test
+  void multiMerge1() {
+    CommitObj a = repo.add(initialCommit());
+    CommitObj b = repo.add(buildCommit("b").addTail(a.id()));
+    CommitObj c = repo.add(buildCommit("c").addTail(a.id()));
+    CommitObj d = repo.add(buildCommit("d").addTail(b.id()));
+    CommitObj e = repo.add(buildCommit("e").addTail(b.id()));
+    CommitObj f = repo.add(buildCommit("f").addTail(c.id()));
+    CommitObj g = repo.add(buildCommit("g").addTail(e.id()).addSecondaryParents(d.id()));
+    CommitObj h = repo.add(buildCommit("h").addTail(d.id()).addSecondaryParents(f.id()));
+    CommitObj i = repo.add(buildCommit("i").addTail(g.id()).addSecondaryParents(h.id()));
+
+    soft.assertThat(
+            MergeBase.builder()
+                .loadCommit(repo::loadCommit)
+                .targetCommitId(f.id())
+                .fromCommitId(i.id())
+                .build()
+                .identifyMergeBase())
+        .isEqualTo(f.id());
+    soft.assertThat(
+            MergeBase.builder()
+                .loadCommit(repo::loadCommit)
+                .respectMergeParents(false)
+                .targetCommitId(f.id())
+                .fromCommitId(i.id())
+                .build()
+                .identifyMergeBase())
+        .isEqualTo(a.id());
+  }
+
+  /**
+   * Merge-again case. <code><pre>
+   * A ----- C - E - F - G
+   *  \- B -/           /
+   *   \    \--------- / ---\
+   *    \- D ---------/------ H - J
+   * </pre></code>
+   *
+   * <p>Best merge-base of {@code J} onto {@code G} is {@code D}.
+   */
+  @Test
+  void multiMerge2() {
+    CommitObj a = repo.add(initialCommit());
+    CommitObj b = repo.add(buildCommit("b").addTail(a.id()));
+    CommitObj c = repo.add(buildCommit("c").addTail(a.id()).addSecondaryParents(b.id()));
+    CommitObj d = repo.add(buildCommit("d").addTail(a.id()));
+    CommitObj e = repo.add(buildCommit("e").addTail(c.id()));
+    CommitObj f = repo.add(buildCommit("f").addTail(e.id()));
+    CommitObj g = repo.add(buildCommit("g").addTail(f.id()).addSecondaryParents(d.id()));
+    CommitObj h = repo.add(buildCommit("h").addTail(d.id()).addSecondaryParents(b.id()));
+    CommitObj j = repo.add(buildCommit("j").addTail(h.id()));
+
+    soft.assertThat(
+            MergeBase.builder()
+                .loadCommit(repo::loadCommit)
+                .targetCommitId(g.id())
+                .fromCommitId(j.id())
+                .build()
+                .identifyMergeBase())
+        .isEqualTo(b.id());
```

Comment:
This assert contradicts the test's javadoc (line 268). I also think `D` is the best merge base in this case.

Reply:
bug in test code, fixed